### ISSUE TITLE
undefined symbol: libpd_print_concatenator

### ISSUE THREAD

**Issue by derekxkwan** [ircam-ismm/node-libpd#3]
Not sure quite what I'm doing wrong but I'm running into a missing symbol libpd_print_concatenator when I load the module like so

`const pd = require('node-libpd');`

I've done the npm install but does this look for any sort of separate libpd install anywhere? i've "make install" ed from the libpd repo and no difference.

If it helps, I'm on Debian Bookworm and node 17.9

**Comment by b-ma** [comment]
Hey,

I just tested on a Mac with node 17.9 and it seems to work properly, so this doesn't seem to be the issue.

I never tested on a Debian system outside Raspberry Pis actually, but there are a few additional step that are not properly documented in the README that you could maybe try

1. Install build tools and some utilities, the node-libpd are not shipped and le package is built on post install (probably `locate` and `sox` are not required here I can't remember exactly what is needed...)
```
sudo apt-get -y install git gcc g++ make locate sox
```

2. Install and build jack2 which is required by the `portaudio` build that is shipped with the library 

```
sudo apt-get install -y libasound2-dev libsndfile1-dev libreadline-dev libreadline6-dev libtinfo-dev libsystemd-dev libsamplerate-dev libdb-dev
git clone https://github.com/jackaudio/jack2.git --depth 1
cd jack2
./waf configure
./waf build
sudo ./waf install
sudo ldconfig
```

Also, an important thing is that you can only use vanilla pd in your patches, as externals should be built with the package

**Comment by b-ma** [comment]
Hum, I think I get the problem... actually only the Node.js bindings are built on `postinstall`, `portaudio` and `libpd` builds are shipped within the library as dynamic libs (see `libpd` and `portaudio` directories in the project). 
Unfortunately, for linux I only included the `portaudio` and `libpd` build for `arm` (i.e. Raspberry Pi) as you can see there : [https://github.com/ircam-ismm/node-libpd/blob/033666ffc9c34dc0e2280b9ecd29a31721cf957b/binding.gyp#L78](https://github.com/ircam-ismm/node-libpd/blob/033666ffc9c34dc0e2280b9ecd29a31721cf957b/binding.gyp#L78), 

So, probably these two should be built for your platform, added to the repo, and registered in the `binding.gyp`file

Do you think you would be confortable enough to do this yourself and make a PR (unfortunately I don't have any Linux machine at hand right now)? I could eventually help digging in my old sources, see how I did that in the beginning.

If not I will try to do it myself, but I probably won't be able to do it before a few weeks

> also, would the vanilla objects include those included with pd/extra such as [bonk~] and [sigmund~]?

Actually, I have no idea (I'm actually very bad using pd...), seems that the objects are properly created in my pd which is quite "raw" I think. Do you have a working patch somewhere?

**Comment by derekxkwan** [comment]
ah, sorry. got busy.

sorry, i'm pretty unfamiliar with the whole module building process in node so i do see the "arm" bits for pi not quite sure what everything does in the gyp file.

sure, just whipped this up. loadbangs starting dsp and [bonk~] is detecting attacks from an [vline~] enveloped [noise~] and spitting out the velocity with [s vel] (so should be able to get floats from the patch every 1000 ms on `vel`)
```

#N canvas 226 121 693 474 12;
#X obj 120 96 metro 1000;
#X msg 120 121 0 \, 1 5 0 \, 0 100 5;
#X obj 120 146 vline~;
#X obj 39 129 noise~;
#X obj 39 154 *~;
#X obj 39 179 bonk~;
#X obj 115 13 loadbang;
#X msg 100 61 1;
#X obj 140 209 list split 2;
#X obj 140 234 list split 1;
#X obj 175 271 s vel;
#X msg 228 51 \; pd dsp 1;
#X obj 119 36 t b b;
#X connect 0 0 1 0;
#X connect 1 0 2 0;
#X connect 2 0 4 1;
#X connect 3 0 4 0;
#X connect 4 0 5 0;
#X connect 5 1 8 0;
#X connect 6 0 12 0;
#X connect 7 0 0 0;
#X connect 8 0 9 0;
#X connect 9 1 10 0;
#X connect 12 0 7 0;
#X connect 12 1 11 0;

```

**Comment by derekxkwan** [comment]
ah, I've managed to get something working by looking at the bindings.gyp file from https://github.com/teropa/node-libpd and adding to the conditions after the aforementioned `OS=="linux"` 

```
 ['target_arch=="x64"', {
                "link_settings": {
                  "libraries": [
                    "<@(module_root_dir)/build/Release/libportaudio.so.2.0.0",
                    "<@(module_root_dir)/build/Release/libpd.so",
                  ],
                  "ldflags": [
                    "-L<@(module_root_dir)/build/Release",
                    "-Wl,-rpath,<@(module_root_dir)/build/Release",
                    '-Wno-sign-compare',
                  ]
                },
                "copies": [
                  {
                    "destination": "build/Release/",
                    "files": [
                      "<@(module_root_dir)/portaudio/lib-x64/libportaudio.so.2.0.0"
                    ]
                  },
                  {
                    "destination": "build/Release/",
                    "files": [
                      "<@(module_root_dir)/libpd/lib-x64/libpd.so"
                    ]
                  }
                ]
              }]

```

and copying over the `libportaudio.so.2.0.0` from apt installing and building my own `libpd.so`. It seems to pass all the tests (with a seg fault at the end, is that expected?) and `[bonk~]` seems to work as well

**Comment by b-ma** [comment]
Hey, cool that's really nice! From your description it seems very correct to me. I just run the tests on my mac and I got the the `Segmentation Fault` as well, I don't really remember if it's expected but it does not seem like regression then.

Could you make a PR with your updates?

**Comment by b-ma** [comment]
released on npm in v1.1.0, thanks again!
